File: 1036-RottingOranges/1036-RottingOranges.js
// Last updated: 4/29/2026, 10:26:15 AM
/**
 * @param {number[][]} grid
 * @return {number}
 */
var orangesRotting = function(grid) {
    const m = grid.length;
    const n = grid[0].length;
    let nFresh = 0;
    let nRotten = 0;
    let queue = [];
    for(let i = 0; i < m; i++) {
        for(let j = 0; j < n; j++) {
            if(grid[i][j] === 1) {
                nFresh++;
            } else if(grid[i][j] === 2) {
                nRotten++;
                queue.push([i, j]);
            }
        }
    }
    console.log(`fresh #: ${nFresh}`)
    console.log(`rotten #: ${nRotten}`)

    if(nFresh === 0) {
        return 0;
    }
    const neighbors = (x, y) => {
        return [[x+1, y], [x-1, y], [x, y+1], [x, y-1]];
    }
    const isValid = (x, y) => {
        return (x >= 0 && x < m && y >= 0 && y < n) && grid[x][y] === 1;
    }
    const visited = new Array(m).fill(null).map(() => new Array(n).fill(0));
    console.log(visited);
    let count = 0;
    while(queue.length !== 0) {
        const nextQueue = [];
        count++;
        for(const [x, y] of queue) {
            visited[x][y] = 1;
            for(const [p, q] of neighbors(x, y)) {
                if(isValid(p, q) && !visited[p][q]) {
                    visited[p][q] = 1;
                    nFresh--;
                    if(nFresh === 0) {
                        return count;
                    }
                    nextQueue.push([p, q]);
                }
            }
        }
        queue = nextQueue;
    }
    return -1;
};

File: 1036-RottingOranges/1036-RottingOranges.py
# Last updated: 4/29/2026, 10:26:19 AM
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        
        visited = [[False]*n for i in range(m)]
        
        queue = []
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    queue.append([i, j])
                    visited[i][j] = True
                if grid[i][j] == 1:
                    count += 1
        
        print(queue)
        print(count)
        
        if count == 0:
            return 0
        
        def valid(i, j):
            return i >= 0 and i < m and j >= 0 and j < n and grid[i][j] == 1
            
        def neighbors(i, j):
            return [[i+1, j], [i-1, j], [i, j+1], [i, j-1]]
        
        t = 0
        while len(queue) != 0:
            next_queue = []
            t += 1
            for i, j in queue:
                next_list = neighbors(i, j)
                for x, y in next_list:
                    if valid(x, y) and visited[x][y] == False:
                        print(f"valid + never visited: {x},{y}")
                        visited[x][y] = True
                        count -= 1
                        if count == 0:
                            return t
                        next_queue.append([x, y])
            queue = next_queue
        return -1
                    